rts/idris_rts.c

```diff
@@ -411,102 +435,193 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
     case BITS8:
-        // max length 8 bit unsigned int str 3 chars (256)
-        cl = allocate(sizeof(Closure) + sizeof(char)*4, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu8, (uint8_t)i->info.bits8);
+        sprintf(temp, "%" PRIu8, (uint8_t)i->info.bits8);
         break;
     case BITS16:
-        // max length 16 bit unsigned int str 5 chars (65,535)
-        cl = allocate(sizeof(Closure) + sizeof(char)*6, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu16, (uint16_t)i->info.bits16);
+        sprintf(temp, "%" PRIu16, (uint16_t)i->info.bits16);
         break;
     case BITS32:
-        // max length 32 bit unsigned int str 10 chars (4,294,967,295)
-        cl = allocate(sizeof(Closure) + sizeof(char)*11, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu32, (uint32_t)i->info.bits32);
+        sprintf(temp, "%" PRIu32, (uint32_t)i->info.bits32);
         break;
     case BITS64:
-        // max length 64 bit unsigned int str 20 chars (18,446,744,073,709,551,615)
-        cl = allocate(sizeof(Closure) + sizeof(char)*21, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu64, (uint64_t)i->info.bits64);
+        sprintf(temp, "%" PRIu64, (uint64_t)i->info.bits64);
         break;
     default:
         fprintf(stderr, "Fatal Error: ClosureType %d, not an integer type", ty);
         exit(EXIT_FAILURE);
     }
 
-    SETTY(cl, STRING);
-    return cl;
+    // avoid invalid UTF-8: all ASCII digits are acceptable as UTF-8 characters.
+    int j;
+    for (j = 0; j < MAX_DIGITS_FOR_64_BITS + 1; j++) {
+        if (temp[j] != '\0' && !(isdigit(temp[j]))) {
+            fprintf(stderr, "Fatal error: invalid output from sprintf on bit type %d!", ty);
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    return MKSTR(vm, (const utf8_byte *)temp);
 }
 
 VAL idris_castStrInt(VM* vm, VAL i) {
     char *end;
-    i_int v = strtol(GETSTR(i), &end, 10);
-    if (*end == '\0' || *end == '\n' || *end == '\r')
-        return MKINT(v);
+    SizedString str = GETSTR(i);
+
+    // First, we need to alloc a C string that's long enough (to get null-termination)
+    char *temp = (char *) malloc(str.size + 1);
+    if (temp == NULL) {
+        fprintf(stderr, "Couldn't allocate temp buffer of size %lu for int->string conversion", str.size);
+        exit(EXIT_FAILURE);
+    }
+    // Then grab the UTF8 bytes out of the Idris string. We don't have
+    // to worry about nulls in the middle because we'll later check to
+    // see if the whole string was used.
+    memcpy(temp, str.str_data, str.size);
+    temp[str.size] = '\0';
+
+    i_int v = strtol(temp, &end, 10);
+    Closure* answer;
+    // If we used the whole string, return the int, otherwise return 0. This is
+    // the meaning of the primitive in the evaluator.
+    if (end == temp + str.size)
```

Comment:
V. small (and possibly paranoia-related and not in keeping with the style guide), but is it beneficial to include braces here, even for the single line? (Here I'm thinking of [#gotofail](https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/).)

```c
if (end == temp + str.size) {
  answer = MKINT(v);
} else {
  answer = MKINT(0);
}